src/KubernetesClient/KubernetesRequest.cs

```diff
@@ -0,0 +1,769 @@
+        /// </summary>
+        public string WatchVersion() => _watchVersion;
+
+        /// <summary>Sets the resource version to use when watching a resource, or empty to watch the current version, or null to not
+        /// execute a watch. The default is null. When set, the response is always streamed (as though <see cref="StreamResponse()"/>
+        /// was true).
+        /// </summary>
+        public KubernetesRequest WatchVersion(string resourceVersion) { _watchVersion = resourceVersion; return this; }
+
+        /// <summary>Adds a value to the query string or headers.</summary>
+        KubernetesRequest Add(ref Dictionary<string,List<string>> dict, string key, string value)
+        {
+            if(string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+            if(dict == null) dict = new Dictionary<string, List<string>>();
+            if(!dict.TryGetValue(key, out List<string> values)) dict[key] = values = new List<string>();
+            values.Add(value);
+            return this;
+        }
+
+        /// <summary>Adds a value to the query string or headers.</summary>
+        KubernetesRequest Add(ref Dictionary<string, List<string>> dict, string key, IEnumerable<string> values)
+        {
+            if(string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+            if(values != null)
+            {
+                if(dict == null) dict = new Dictionary<string, List<string>>();
+                if(!dict.TryGetValue(key, out List<string> list)) dict[key] = list = new List<string>();
+                list.AddRange(values);
+            }
+            return this;
+        }
+
+        /// <summary>Sets a value in the query string or headers.</summary>
+        KubernetesRequest Set(ref Dictionary<string,List<string>> dict, string key, string value)
+        {
+            if(string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+            dict = dict ?? new Dictionary<string, List<string>>();
+            if(!dict.TryGetValue(key, out List<string> values)) dict[key] = values = new List<string>();
+            values.Clear();
+            values.Add(value);
+            return this;
+        }
+
+        /// <summary>Creates an <see cref="HttpRequestMessage"/> representing the current request.</summary>
+        HttpRequestMessage CreateRequestMessage()
+        {
+            var req = new HttpRequestMessage(Method(), GetRequestUri());
+
+            // add the headers
+            if(!string.IsNullOrEmpty(config?.AccessToken))
+            {
+                req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", config.AccessToken);
+            }
+            else if(!string.IsNullOrEmpty(config?.Username))
+            {
+                req.Headers.Authorization = new AuthenticationHeaderValue("Basic",
+                    Convert.ToBase64String(Encoding.UTF8.GetBytes(config.Username + ":" + config.Password)));
+            }
```

Comment:
I think currently it's not handled by message handlers attached to the HttpClient even in the normal Kubernetes client. But it does construct a Credentials object rather than setting the header manually (although the effect is ultimately the same). I think I'll change this to use the same Credentials object that the client has already constructed.

I saw your PR for refreshing the access token, by the way. Seems good and important to have.

Reply:
I changed this to

    if(credentials != null) await credentials.ProcessHttpRequestAsync(req, cancelToken);

where credentials is taken from the Kubernetes object or supplied by the user. It'll be in the next iteration.